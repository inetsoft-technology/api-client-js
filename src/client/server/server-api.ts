/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * ServerResourceUsage contains information about the server resource usage on a host at a
 * particular point in time.
 */
export interface ServerResourceUsage {
   /**
    * The ISO 8601-formatted date and time at which the resource usage was recorded.
    */
   timestamp: string;

   /**
    * The hostname of the machine from which the resource usage was recorded.
    */
   host: string;

   /**
    * The CPU usage as a percentage.
    */
   cpuUsage: number;

   /**
    * The memory usage in bytes.
    */
   memoryUsage: number;

   /**
    * The number of times that the garbage collector has run during the last sampling period.
    */
   gcCount: number;

   /**
    * The total amount of time, in milliseconds, that was spent in garbage collection during the
    * last sampling interval.
    */
   gcTime: number;

   /**
    * The number of reports that were executed during the last sampling interval.
    */
   executingReports: number;

   /**
    * The number of viewsheets that were executed during the last sampling interval.
    */
   executingViewsheets: number;

   /**
    * The number of queries that were executed during the last sampling interval.
    */
   executingQueries: number;
}

/**
 * ServerApi provides operations for getting information about the servers.
 */
export interface ServerApi {
   /**
    * Gets the server usage history.
    *
    * @return a list of ServerResourceUsage structures with data sampled at a regular interval.
    */
   getUsageHistory(): Promise<ServerResourceUsage[]>;
}
