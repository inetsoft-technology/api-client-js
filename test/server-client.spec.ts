/*
 * Copyright 2019 InetSoft Technology
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ServerClient } from "../src/client/server/server-client";

describe("ServerClient", () => {
   let adapter: any;
   let client: ServerClient;

   const history = [
      {
         timestamp: "2019-07-26T09:00:00Z",
         host: "localhost",
         cpuUsage: 50.8,
         memoryUsage: 104857600,
         gcCount: 1,
         gcTime: 250,
         executingReports: 5,
         executingViewsheets: 4,
         executingQueries: 10
      },
      {
         timestamp: "2019-07-26T09:00:01Z",
         host: "localhost",
         cpuUsage: 25.2,
         memoryUsage: 104857610,
         gcCount: 1,
         gcTime: 200,
         executingReports: 1,
         executingViewsheets: 2,
         executingQueries: 4
      },
   ];

   beforeEach(() => {
      adapter = {
         get: jest.fn(),
         post: jest.fn(),
         put: jest.fn(),
         delete: jest.fn(),
         map: jest.fn(),
         mapText: jest.fn(),
         mapBinary: jest.fn()
      };
      client = new ServerClient(adapter, () => Promise.resolve("secret"), "http://localhost:8080/sree/");
   });

   describe("when getUsageHistory() is called", () => {
      it("the a list of server usage structures is returned", () => {
         adapter.get.mockReturnValueOnce(Promise.resolve({data: {history}}));
         adapter.map.mockImplementation((promise: Promise<any>) => promise.then(response => response.data));
         return client.getUsageHistory().then(result => {
            expect(result).toMatchSnapshot();
            expect(adapter.get.mock.calls.length).toBe(1);
            expect(adapter.get.mock.calls[0][0]).toBe("http://localhost:8080/sree/api/public/server/usage");
         });
      });
   });
});
