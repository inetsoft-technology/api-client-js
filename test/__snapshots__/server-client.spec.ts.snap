// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerClient when getUsageHistory() is called the a list of server usage structures is returned 1`] = `
Array [
  Object {
    "cpuUsage": 50.8,
    "executingQueries": 10,
    "executingReports": 5,
    "executingViewsheets": 4,
    "gcCount": 1,
    "gcTime": 250,
    "host": "localhost",
    "memoryUsage": 104857600,
    "timestamp": "2019-07-26T09:00:00Z",
  },
  Object {
    "cpuUsage": 25.2,
    "executingQueries": 4,
    "executingReports": 1,
    "executingViewsheets": 2,
    "gcCount": 1,
    "gcTime": 200,
    "host": "localhost",
    "memoryUsage": 104857610,
    "timestamp": "2019-07-26T09:00:01Z",
  },
]
`;
